RecoMuon/L2MuonSeedGenerator/src/L2MuonSeedGeneratorFromL1T.cc

```diff
@@ -124,246 +157,804 @@ void L2MuonSeedGeneratorFromL1T::produce(edm::Event& iEvent, const edm::EventSet
+      unsigned int NofflineSeed = offlineSeedHandle->size();
+      LogTrace(metname) << "Number of offline seeds " << NofflineSeed << endl;
+
+      // Initialize dRmtx and selOffseeds
+      dRmtx = vector< vector<double> >(NmuColl, vector<double>(NofflineSeed, 999.0));
+      selOffseeds = vector< vector <const TrajectorySeed *> >(NmuColl, vector<const TrajectorySeed *>(NofflineSeed, nullptr));
+    }
+
+    // At least one L1 muons are associated with offSeed
+    bool isAsso = false;
+
+    //--- Fill dR matrix
+    for (int ibx = muColl->getFirstBX(); ibx <= muColl->getLastBX(); ++ibx) {
+      if (centralBxOnly_ && (ibx != 0)) continue;
+      for (auto it = muColl->begin(ibx); it != muColl->end(ibx); it++){
+
+        //Position of given L1mu
+        unsigned int imu = distance(muColl->begin(muColl->getFirstBX()),it);
+
+        unsigned int quality = it->hwQual();
+        int valid_charge = it->hwChargeValid();
+
+        float pt    =  it->pt();
+        float eta   =  it->eta();
+        float theta =  2*atan(exp(-eta));
+        float phi   =  it->phi();
+        int charge  =  it->charge();
+        // Set charge=0 for the time being if the valid charge bit is zero
+        if (!valid_charge) charge = 0;
+
+        int link = 36 + (int)(it -> tfMuonIndex() / 3.);
+        bool barrel = true;
+        if ( (link >= 36 && link <= 41) || (link >= 66 && link <= 71)) barrel = false;
+
+        if ( pt < theL1MinPt || fabs(eta) > theL1MaxEta ) continue;
+
+        LogTrace(metname) << "New L2 Muon Seed" ;
+        LogTrace(metname) << "Pt = "         << pt     << " GeV/c";
+        LogTrace(metname) << "eta = "        << eta;
+        LogTrace(metname) << "theta = "      << theta  << " rad";
+        LogTrace(metname) << "phi = "        << phi    << " rad";
+        LogTrace(metname) << "charge = "     << charge;
+        LogTrace(metname) << "In Barrel? = " << barrel;
+
+        if ( quality <= theL1MinQuality ) continue;
+        LogTrace(metname) << "quality = "<< quality;
+
+        // Update the services
+        theService->update(iSetup);
+
+        const DetLayer *detLayer = nullptr;
+        float radius = 0.;
+
+        CLHEP::Hep3Vector vec(0.,1.,0.);
+        vec.setTheta(theta);
+        vec.setPhi(phi);
+
+        DetId theid;
+        // Get the det layer on which the state should be put
+        if ( barrel ){
+          LogTrace(metname) << "The seed is in the barrel";
+
+          // MB2
+          DetId id = DTChamberId(0,2,0);
+          detLayer = theService->detLayerGeometry()->idToLayer(id);
+          LogTrace(metname) << "L2 Layer: " << debug.dumpLayer(detLayer);
+
+          const BoundSurface* sur = &(detLayer->surface());
+          const BoundCylinder* bc = dynamic_cast<const BoundCylinder*>(sur);
+
+          radius = fabs(bc->radius()/sin(theta));
+          theid  = id;
+
+          LogTrace(metname) << "radius "<<radius;
+
+          if ( pt < 3.5 ) pt = 3.5;
```

Comment:
What is special with "3.5"?
Shouldn't this threshold better made parameterizable, or taken from somewhere else?